### PR TITLE
improve(liquidator): Ignore WDF if no withdrawal request to extend

### LINKED ISSUE #2510: Liquidator WDF strategy: Consider whether bot should ignore WDF constraint if no withdrawal request to extend
**Why**

If withdrawal request has expired or does not exist, consider whether bots should still maintain the WDF balance invariant. For instance, should bot skip liquidating an undercollateralized position if it has <= WDF in its balance?

**How**

Add a check to the `liquidationStrategy.js` file's `calculateTokensToLiquidate()` function to check if a withdrawal request exists.

** Difficulty **
- 3: ~1 hour

### PR DESCRIPTION
**Pull request by nicholaspai**
**Motivation**

#2510 

Related to #2508

**Summary**

Adds check to `liquidationStrategy` to ignore WDF threshold if no withdrawal request to extend. The WDF is typically used to maintain a balance of funds with which to extend withdrawal timers, but this should be ignored if there are no withdrawals to process.

**However, this is a controversial PR because merging it would mean that we can no longer guarantee that a bot will always have >= WDF in token balance.**

Something to consider if we merge this PR: Does this render the `whaleDefenseFundWei` parameter irrelevant? I.e. can we merge this PR AND subsequently remove that param?

**Testing**

Check a box to describe how you tested these changes and list the steps for reviewers to test.

- [ ]  Ran end-to-end test, running the code as in production
- [X]  New unit tests created
- [ ]  Existing tests adequate, no new tests required
- [X]  All existing tests pass
- [ ]  Untested

**Issue(s)**

Fixes #2510

### CODE REVIEW

File: packages/liquidator/src/liquidationStrategy.js [hunk 1]
```diff
@@ -90,17 +90,27 @@ module.exports = (
...
 
   // use in the case of a non whale defense liquidation
-  // Similar to the logic that already exists, but takes into account more information
-  function calculateTokensToLiquidate({ syntheticTokenBalance, positionTokens, maxTokensToLiquidateWei }) {
+  // Similar to the logic that already exists, but takes into account more information.
+  function calculateTokensToLiquidate({
+    syntheticTokenBalance,
+    positionTokens,
+    maxTokensToLiquidateWei,
+    withdrawRequestPending
```
Comment (chrismaree) [inline packages/liquidator/src/liquidationStrategy.js:98]: nice to give the _option_ to keep the current bot behavior.

File: packages/liquidator/src/liquidationStrategy.js [hunk 2]
```diff
@@ -197,13 +208,18 @@ module.exports = (
     return 100 * (elapsed / duration);
   }
 
-  // Returns true if enough of the withdrawal liveness has passed that the WDF strategy can be activated.
-  function passedDefenseActivationPercent({ position, currentBlockTime }) {
+  function hasWithdrawRequestPending({ position, currentBlockTime }) {
```
Comment (chrismaree) [inline packages/liquidator/src/liquidationStrategy.js:211]: +1

**Review by chrismaree — APPROVED** [review]
image

File: packages/liquidator/mocha-test/liquidationStrategy.js [hunk 3]
```diff
@@ -81,9 +81,10 @@ describe("LiquidatorStrategy", () => {
     let result = strat.utils.calculateTokensToLiquidate({
       syntheticTokenBalance: 100,
       positionTokens: 10,
-      maxTokensToLiquidateWei: "1000"
+      maxTokensToLiquidateWei: "1000",
+      withdrawRequestPending: true
     });
-    // should respsect position size
```
[the remark this answers is not in the document]
Comment (mrice32) [reply packages/liquidator/mocha-test/liquidationStrategy.js:86]: repsect = 🐍  + 🐞

**Review by mrice32 — APPROVED** [review]
LGTM! This is pretty subtle. +1

File: packages/liquidator/mocha-test/liquidationStrategy.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (daywiss) [reply packages/liquidator/mocha-test/liquidationStrategy.js:86]: i just got that

**Review by daywiss — APPROVED** [review]
nice update

**Pull request merged**